src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -10030,11 +10061,19 @@ private bool ScanParenthesizedImplicitlyTypedLambda(Precedence precedence)
             }
 
             //  case 2:  ( x ) =>
-            if (IsTrueIdentifier(this.PeekToken(1))
-                && this.PeekToken(2).Kind == SyntaxKind.CloseParenToken
-                && this.PeekToken(3).Kind == SyntaxKind.EqualsGreaterThanToken)
+            if (IsTrueIdentifier(this.PeekToken(1)))
             {
-                return true;
+                if (this.PeekToken(2).Kind == SyntaxKind.CloseParenToken
+                && this.PeekToken(3).Kind == SyntaxKind.EqualsGreaterThanToken)
+                {
+                    return true;
+                }
+                if (this.PeekToken(2).Kind == SyntaxKind.ExclamationToken
+                    && this.PeekToken(3).Kind == SyntaxKind.CloseParenToken
+                    && this.PeekToken(4).Kind == SyntaxKind.EqualsGreaterThanToken)
+                {
+                    return true;
+                }
```

Comment:
another way to writ this (which i think is cleaner is this):

```c#

// allow for       a) =>      or     a!) =>
var skipIndex = 2;
if (PeekToken(skipIndex).Kind == Exclamation))
{
    skipIndex++;
}

// Must have:     ) => 
if (this.PeekToken(skipIndex + 1).Kind == SyntaxKind.CloseParenToken
    && this.PeekToken(skipIndex + 2).Kind == SyntaxKind.EqualsGreaterThanToken)
{
    return true;
}
```

This way we keep teh logic the same for the part we care about (recognizing the `=>`), but we basically make it optional as to if we see a `!` or not.

--

Note: this is a suggestion.  If you prefer the approach you've taken, then feel free to stick with it :)

This way we just simply